packages/cosmic-swingset/lib/ag-solo/vats/lib-dehydrate.js

```diff
@@ -0,0 +1,82 @@
+import harden from '@agoric/harden';
+import { makeMarshal } from '@agoric/marshal';
+import makeStore from '@agoric/store';
+import { assert, details } from '@agoric/assert';
+
+// Marshalling for the UI should only use the user's petnames. We will
+// call marshalling for the UI "dehydration" and "hydration" to distinguish it from
+// other marshalling.
+export const makeDehydrator = (initialUnnamedCount = 0) => {
+  let unnamedCount = initialUnnamedCount;
+
+  const petnameKindToMappings = makeStore();
+
+  const searchOrder = [];
+
+  const makeMappings = kind => {
+    assert.typeof(kind, 'string', details`kind ${kind} must be a string`);
+    const valToPetname = makeStore();
+    const petnameToVal = makeStore();
+    const addPetname = (petname, val) => {
+      petnameToVal.init(petname, val);
+      valToPetname.init(val, petname);
+    };
+    const mappings = harden({
+      valToPetname,
+      petnameToVal,
+      addPetname,
+      kind,
+    });
+    petnameKindToMappings.init(kind, mappings);
+    return mappings;
+  };
+
+  const unnamedMappings = makeMappings('unnamed');
+
+  const addToUnnamed = val => {
+    unnamedCount += 1;
+    const placeholder = `unnamed-${unnamedCount}`;
+    const placeholderName = harden({
+      kind: 'unnamed',
+      petname: placeholder,
+    });
+    unnamedMappings.addPetname(placeholder, val);
+    return placeholderName;
+  };
+
+  // look through the petname stores in order and create a new
+  // unnamed record if not found.
+  const convertValToName = val => {
+    for (let i = 0; i < searchOrder.length; i += 1) {
+      const kind = searchOrder[i];
+      const { valToPetname } = petnameKindToMappings.get(kind);
+      if (valToPetname.has(val)) {
+        return harden({
+          kind,
+          petname: valToPetname.get(val),
+        });
+      }
+    }
+    // not found in any map
+    const placeholderName = addToUnnamed(val);
+    return placeholderName;
+  };
+
+  const convertNameToVal = ({ kind, petname }) => {
+    const { petnameToVal } = petnameKindToMappings.get(kind);
+    return petnameToVal.get(petname);
+  };
+  const { serialize: dehydrate, unserialize: hydrate } = makeMarshal(
+    convertValToName,
+    convertNameToVal,
+  );
+  return harden({
+    hydrate,
+    dehydrate,
```

Comment:
So I'm worried about spoofing attacks, but I can't tell how this is meant to be used, so I'm not sure if the attacker has enough access to mount one.

`marshal`'s serialization looks for pass-by-reference objects and lets the callback transform them into some specially-marked copyable-data representation. The deserialization looks for those specially-marked objects and passes them to the other callback for transformation back into something else. In our usual case, that's a Presence, or the original pass-by-reference object. That object is not representable in the data form: the transformation takes place between two different domains. No matter what the caller passes as the `body` of the `args` capdata structure, they cannot cause `deserialize` to produce the same kind of Presence that they could get by using the normal reference syntax.

But here, the Presence equivalent is just plain data. The attacker could put `{ kind, name }` in the data they send to you, and the deserialized output would be indistinguishable from what you would have gotten if they'd put the object you were intending to translate. This might trick you into thinking you were interacting with an object that you'd previously assigned a name to, but was actually just plain data.

Now, if the use case of this is to take a Presence (which you got as the normal arguments inside a vat, i.e. it's passed through the regular `unserialize`, and which you've confirmed is really a Presence somehow), and do a lookup with `dehydrate(presence)` to see if you have a name for it, then I think that's safe. The attacker can't trick you into using `{ kind, name }` as the input to `dehydrate` because you're testing that it's a real Presence.

How is this meant to be used? I'm wondering if a safer API might be to translate one Presence at a time, rather than taking the shape of a general-purpose serialize/deserialize tool. Such an API could assert that its argument was a Presence before doing anything else. It wouldn't need `makeMarshal`, and the API might be more like `{ name, kind } = getNameForPresence(p)` and `setNameForPresence(p, name, kind)` or something.

Reply:
Thanks for the analysis @warner. I want to provide some context and make sure I understand the spoofing attack correctly. 

This code will be used to in communication between the wallet backend (which has presences) and the wallet frontend (which does not have presences). We control both of these. The code will also be used in communication between the wallet backend and dapp frontends, although we are still determining how much actual access the dapp frontends will get. We do not control the dapp frontend, although @mfig has some great ideas about how to embed information from the wallet into the dapp UI without the dapp UI actually having any access to it. So, I think we should analyze this from the perspective of the dapp UI being the potential attacker, at least in the short-term.

Anything being sent *to* the wallet UI or dapp UI will be serialized/dehydrated. Anything received *from* the wallet UI or dapp UI will be deserialized/hydrated.

So if I understand the attempted spoofing attack correctly, the dapp UI would pass serialized/dehydrated data to the wallet backend, for instance: 
```
harden({
    body: '{"handle":{"@qclass":"slot","index":0}}',
    slots: [{ kind: 'instanceHandle', petname: 'autoswap' }],
})
```

But, the wallet backend will always attempt to deserialize/hydrate what it receives from the UI, so this will correctly be turned into a real handle. 

As far as I can tell, at worst, the UI can guess at petnames that it hasn't previously been given access to, since the petnames are of course human-readable and not in any way random. We should be able to control this though. 

Can you walk me through the spoofing attack in this context? I'm not seeing it here.

Reply:
I think the attack would be where the dapp UI sends:

```
harden({
    body: '{"handle":{"kind":"instanceHandle","petname":"autoswap"}}',
    slots: [],
})
```

Both that serialized data, and the one you wrote, will hydrate into the same thing:

```
{ kind: 'instanceHandle', petname: 'autoswap' }
```

(the mapping is "non-injective", if I have that right)

If the attacker gets to supply the serialized data, they have full control over what comes out of hydration, and this non-injective-ness is boring. But if the attacker is limited to feeding data *into* a serializer (and the serializer owns a petname table that the attacker doesn't get to control), then this non-injective-ness gives the attacker the ability to make the deserializer emit petnames that would otherwise be unreachable.

I still won't claim to understand how this is being used, so maybe this isn't a real problem, but petname/lambda-name systems are usually about reliably/unspoofably identifying incoming references as ones you've seen before, so any place where that might be spoofable deserves extra attention.

Reply:
No, that's incorrect. The "correct" serialized form and the "spoofing attack" serialized form will hydrate to two very different things. 

1. The correct serialized form:

```
harden({
    body: '{"handle":{"@qclass":"slot","index":0}}',
    slots: [{ kind: 'instanceHandle', petname: 'autoswap' }],
})
```
will deserialize/hydrate to:

```
{ handle: handle4 },
```
(This result has been viewable as a test in this PR.)

2. The spoofing attack that you provided:

```
harden({
    body: '{"handle":{"kind":"instanceHandle","petname":"autoswap"}}',
    slots: [],
}),
```
will deserialize/hydrate to:

```
{ handle: { kind: 'instanceHandle', petname: 'autoswap' } }
```

These are not the same thing. There's no non-injective-ness.

Additionally, the scenario you mention (the attacker using a serializer/dehydrating) isn't a potential scenario. As I explained above, the only likely attacker is the dapp UI sending messages to the wallet backend. (The dapp UI has no presences and no access to either the `hydrate` or `dehydrate` functions.) The wallet backend will assume that the data from the dapp UI is serialized (indeed, any presences can't even be transferred this way if anyone wanted to). The wallet backend will attempt to *deserialize/hydrate* the message from the dapp UI, to turn the data and petname slots into real objects. This is the opposite of the scenario you mention.

Again, like I said above, the only potential attack I can see is that the dapp UI could guess at petnames (because they are inherently guessable) and without any other protection mechanisms, the wallet could hydrate to an object that it didn't expect the dapp UI to know about. That seems easily fixable though, by recording what the dapp UI has been told about or should have access to. That task will be in a different PR in which this library is applied to the wallet.

I think I've put good effort into explaining what the actual scenario is and how the dehydrator works, and I think I've shown that this particular concern on examination does not apply to this PR, so I'd like to merge.

Reply:
ok, that sounds good.. just wanted to make sure we'd thought through the issue. thanks for the detailed investigation!